README
NAME
    Config::Auto - Magical config file parser

SYNOPSIS
      use Config::Auto;

      # Not very magical at all.
      my $config = Config::Auto::parse("myprogram.conf", format => "colon");

      # Considerably more magical.
      my $config = Config::Auto::parse("myprogram.conf");

      # Highly magical.
      my $config = Config::Auto::parse();

DESCRIPTION
    This module was written after having to write Yet Another Config File
    Parser for some variety of colon-separated config. I decided "never
    again".

    When you call "Config::Auto::parse" with no arguments, we first look at
    $0 to determine the program's name. Let's assume that's "snerk". We look
    for the following files:

        snerkconfig
        ~/snerkconfig
        /etc/snerkconfig
        snerk.config
        ~/snerk.config
        /etc/snerk.config
        snerkrc
        ~/snerkrc
        /etc/snerkrc
        .snerkrc
        ~/.snerkrc
        /etc/.snerkrc

    We take the first one we find, and examine it to determine what format
    it's in. The algorithm used is a heuristic "which is a fancy way of
    saying that it doesn't work." (Mark Dominus.) We know about colon
    separated, space separated, equals separated, XML, Perl code, Windows
    INI, BIND9 and irssi style config files. If it chooses the wrong one,
    you can force it with the "format" option.

    If you don't want it ever to detect and execute config files which are
    made up of Perl code, set "$Config::Auto::DisablePerl = 1".

    Then the file is parsed and a data structure is returned. Since we're
    working magic, we have to do the best we can under the circumstances -
    "You rush a miracle man, you get rotten miracles." (Miracle Max) So
    there are no guarantees about the structure that's returned. If you have
    a fairly regular config file format, you'll get a regular data structure
    back. If your config file is confusing, so will the return structure be.
    Isn't life tragic?

    Here's what we make of some common Unix config files:

    /etc/resolv.conf:

        $VAR1 = {
              'nameserver' => [ '163.1.2.1', '129.67.1.1', '129.67.1.180' ],
              'search' => [ 'oucs.ox.ac.uk', 'ox.ac.uk' ]
            };

    /etc/passwd:

        $VAR1 = {
              'root' => [ 'x', '0', '0', 'root', '/root', '/bin/bash' ],
              ...
            };

    /etc/gpm.conf:

        $VAR1 = {
              'append' => '""',
              'responsiveness' => '',
              'device' => '/dev/psaux',
              'type' => 'ps2',
              'repeat_type' => 'ms3'
            };

    /etc/nsswitch.conf:

        $VAR1 = {
              'netgroup' => 'nis',
              'passwd' => 'compat',
              'hosts' => [ 'files', 'dns' ],
              ...
        };

TODO
    Better BIND9 file format parser.
    irssi file format parser currently doesn't exist. It would be
    good to add support for "mutt" and "vim" style "set"-based RCs.

AUTHOR
    Simon Cozens, "simon@cpan.org"

LICENSE
    AL&GPL.



Changes
0.17    Wed Feb 15 15:02:00 2006
================================
    - Added minimal support for bind style config files

0.15_01 Thu Aug 11 17:00:50 2005
================================
    - Added experimental YAML support

0.14    Thu Aug 11 16:27:50 2005
================================
    - Fix uninitialized value warning
    - Fix pod error
    - Files under /etc we're not properly detected
    - Fix casing of 'perl' in the Formats section of the pod
    - Make formats be accepted in any casing


0.13    Tue Jul 26 16:35:37 2005
================================
    - Add optional path support
    - update test suite to test::more
    - split tests by different formats
    - Patch supplied by ssoriche@coloredblocks.net, tweaked and applied


0.12    Thu Mar 10 14:58:06 2005
================================
    - extend do() diagnostics for perl files
    - extend documentation

0.11    Wed Nov 10 11:35:41 2004
================================
    - fix small thinko in file finding logic.

0.10    Tue Aug 17 14:34:41 2004
================================
    - added support for fixed config file name.

0.07 ???
===================================
    - added support for '/usr/local/etc/'

0.06    Sat Feb 21 13:21:43 2004
===================================
    - fix a bug in tr/// on a string

0.05    Tue Feb 10 13:16:59 2004
===================================
    - Fix so that a config file is magically found when the program
        is started with a full pathname as well

0.04    Sun Sep  7 19:28:40 2003
====================================
    - Jos Boumans took maintainership
    - XML::Simple is now an optional module

0.03    Sun Aug  4 09:51:42 2002
====================================
    - INI files patch by Garrett Rooney

0.02    Thu Jul  4 17:20:13 2002
====================================
    - I always screw up the README.

====================================
0.01  Wed Jul  3 15:00:16 2002
	- original version; created by h2xs 1.21 with options
		-AX -n Config::Auto



Makefile.PL
use ExtUtils::MakeMaker;

use strict;
use Getopt::Std;
use Data::Dumper;
my $opts = {};
getopts( 'x', $opts );

my $have_xml_simple = eval { require XML::Simple; 1; };

if( !$have_xml_simple and !$opts->{'x'} ) {
    warn qq[Since the 0.04 release, XML::Simple is an optional prerequisite.\n].
         qq[If you'd like to install Config::Auto with XML support, please\n] .
         qq[rerun this Makefile.PL with the '-x' option\n];
}

my $prereqs = {
    'XML::Simple'       => 0,
    'YAML'              => 0,
    'Config::IniFiles'  => 0,
};
delete $prereqs->{'XML::Simple'} unless $opts->{'x'};             

WriteMakefile(
    'NAME'		        => 'Config::Auto',
    'VERSION_FROM'	    => 'lib/Config/Auto.pm', # finds $VERSION
    'PREREQ_PM'		    => $prereqs, 
    ( $] >= 5.005 
        ? ( ABSTRACT_FROM    => 'lib/Config/Auto.pm', # retrieve abstract from module
            AUTHOR           => 'Jos I. Boumans <kane@cpan.org>') 
        : ()
    ),
);


lib/Config/Auto.pm
package Config::Auto;

use strict;
use warnings;
use File::Spec::Functions;
use File::Basename;
#use XML::Simple;   # this is now optional
use Config::IniFiles;
use Carp;

use vars qw[$VERSION $DisablePerl $Untaint $Format];

$VERSION = '0.17';
$DisablePerl = 0;
$Untaint = 0;

my %methods = (
    perl   => \&eval_perl,
    colon  => \&colon_sep,
    space  => \&space_sep,
    equal  => \&equal_sep,
    bind   => \&bind_style,
    irssi  => \&irssi_style,
    xml    => \&parse_xml,
    ini    => \&parse_ini,
    list   => \&return_list,
    yaml   => \&yaml,
);

delete $methods{'xml'}
    unless eval { require XML::Simple; XML::Simple->import; 1 };

sub parse {
    my $file = shift;
    my %args = @_;

    $file = find_file($file,$args{path})    if not defined $file or 
                                                not -e $file;
    croak "No config file found!"           if not defined $file;
    croak "Config file $file not readable!" if not -e $file;

    return if -B $file;

    my $method;
    my @data;

    if (!defined $args{format}) {
        # OK, let's take a look at you.
        my @data;
        open CONFIG, $file or croak "$file: $!";
        if (-s $file > 1024*100) {
            # Just read in a bit.
            while (<CONFIG>) {
                push @data, $_;
                last if $. >= 50;
            }
        } else {
            @data = <CONFIG>;
        }
        my %scores = score(\@data);
        delete $scores{perl} if exists $scores{perl} and $DisablePerl;
        croak "Unparsable file format!" if !keys %scores;
        # Clear winner?
        my @methods = sort { $scores{$b} <=> $scores{$a} } keys %scores;
        if (@methods > 1) {
            croak "File format unclear! ".join ",", map { "$_ => $scores{$_}"} @methods
               if $scores{$methods[0]} == $scores{$methods[1]};
        }
        $method = $methods[0];
    } else {
        croak "Unknown format $args{format}: use one of @{[ keys %methods ]}"
            if not exists $methods{ lc $args{format} };
        $method = lc $args{format};
    }

    $Format = $method;
    return $methods{$method}->($file);
}

sub score {
    my $data_r = shift;
    return (xml => 100)     if $data_r->[0] =~ /^\s*<\?xml/;
    return (perl => 100)    if $data_r->[0] =~ /^#!.*perl/;
    my %score;

    for (@$data_r) {
        ### it's almost definately YAML if the first line matches this
        $score{yaml} += 20              if /--- #YAML:/ and $data_r->[0] eq $_;
    
        # Easy to comment out foo=bar syntax
        $score{equal}++                 if /^\s*#\s*\w+\s*=/;
        next if /^\s*#/;

        $score{xml}++                   for /(<\w+.*?>)/g;
        $score{xml}+= 2                 for m|(</\w+.*?>)|g;
        $score{xml}+= 5                 for m|(/>)|g;
        next unless /\S/;

        $score{equal}++, $score{ini}++  if m|^.*=.*$|;
        $score{equal}++, $score{ini}++  if m|^\S+\s+=\s+|;
        $score{colon}++                 if /^[^:]+:[^:=]+/;
        $score{colon}+=2                if /^\s*\w+\s*:[^:]+$/;
        $score{colonequal}+= 3          if /^\s*\w+\s*:=[^:]+$/; # Debian foo.
        $score{perl}+= 10               if /^\s*\$\w+(\{.*?\})*\s*=.*/;
        $score{space}++                 if m|^[^\s:]+\s+\S+$|;

        # mtab, fstab, etc.
        $score{space}++                 if m|^(\S+)\s+(\S+\s*)+|;
        $score{bind}+= 5                if /\s*\S+\s*{$/;
        $score{list}++                  if /^[\w\/\-\+]+$/;
        $score{bind}+= 5                if /^\s*}\s*$/  and exists $score{bind};
        $score{irssi}+= 5               if /^\s*};\s*$/ and exists $score{irssi};
        $score{irssi}+= 10              if /(\s*|^)\w+\s*=\s*{/;
        $score{perl}++                  if /\b([@%\$]\w+)/g;
        $score{perl}+= 2                if /;\s*$/;
        $score{perl}+=10                if /(if|for|while|until|unless)\s*\(/;
        $score{perl}++                  for /([\{\}])/g;
        $score{equal}++, $score{ini}++  if m|^\s*\w+\s*=.*$|;
        $score{ini} += 10               if /^\s*\[[\s\w]+\]\s*$/;
    }

    # Choose between Win INI format and foo = bar
    if (exists $score{ini}) {
        $score{ini} > $score{equal}
            ? delete $score{equal}
            : delete $score{ini};
    }

    # Some general sanity checks
    if (exists $score{perl}) {
        $score{perl} /= 2   unless ("@$data_r" =~ /;/) > 3 or $#$data_r < 3;
        delete $score{perl} unless ("@$data_r" =~ /;/);
        delete $score{perl} unless ("@$data_r" =~ /([\$\@\%]\w+)/);
    }

    return %score;
}

sub find_file {
    my($file,$path) = @_;

    my $x;
    my $whoami = basename($0);
    my $bindir = dirname($0);

   $whoami =~ s/\.(pl|t)$//;
 
   my @filenames = $file ||
                     ("${whoami}config", "${whoami}.config", 
                      "${whoami}rc", ".${whoami}rc");


   foreach my $filename (@filenames) {
       return $filename        if -e $filename;
       return $x               if ($x = _chkpaths($path,$filename)) and -e $x;
       return $x               if -e ($x = catfile($bindir,$filename));
       return $x               if -e ($x = catfile($ENV{HOME},$filename));
       return "/etc/$filename" if -e "/etc/$filename";
       return "/usr/local/etc/$filename"  
                               if -e "/usr/local/etc/$filename";
    }
    return undef;
}

sub _chkpaths {
    my ($paths,$filename)=@_;
    my $file;

    if ($paths) {

        if(ref($paths) eq 'ARRAY') {
            foreach my $path (@$paths) {
                return $file if -e ($file = catfile($path,$filename));
            }
            
        } else {
            return $file     if -e ($file = catfile($paths,$filename));
        }
    
    } else {
        return undef;
    }
}

sub eval_perl   {
  my $file = shift;
  ($file) = $file =~ m/^(.*)$/s if $Untaint;
  my $cfg = do $file;
  croak __PACKAGE__ . " couldn't parse $file: $@" if $@;
  croak __PACKAGE__ . " couldn't do $file: $!"    unless defined $cfg;
  croak __PACKAGE__ . " couldn't run $file"       unless $cfg;
  return $cfg;
}

sub parse_xml   { return XMLin(shift); }
sub parse_ini   { tie my %ini, 'Config::IniFiles', (-file=>$_[0]); 
                    return \%ini; }
sub return_list { open my $fh, shift or die $!; return [<$fh>]; }
sub yaml        { require YAML; return YAML::LoadFile( shift ) }

#
# Parse bind-style config files, should work with bind versions 8/9.
# Parser is far from perfect. Still doesn't parse single line blocks.
#
sub bind_style {
    my $file = $_[0];
    open IN, $file or die("Can't open file $file: $!");
    my %config;

    # Define parser stacks for block names and block options
    my $block_name = '';
    my $block_opts = \%config;
    my @block_stack = ( $block_name );
    my @block_opt_stack = ( $block_opts );
    my @values;

    #
    # Compile regex objects for parsing
    # This is a *simple* parsing that for now is not supposed to work
    # with every single bind style config file.
    #
    my %rx = (
        blank         => qr/^\s*(#|\/\/)/,                  # (either in sh or c++ style)
        include       => qr/^\s*include\s*"(.*)";\s*/,
        block_opening => qr/^\s*(.+)\s*{\s*/,
        block_closing => qr/^\s*}\s*;?\s*/,
        option        => qr/^\s*([^\s]+)\s*(.+)\s*;\s*/,
        value         => qr/^\s*([^\s]+)\s*;\s*/,
        zone          => qr/zone\s*"?([^\s"]+)"?\s*/,
    );

    while (<IN>)
    {

        #warn('   --- ' . $_);

        # Skip blanks and comments
        next if $_ =~ $rx{blank};

        # Parse "include" directives
        # For now we don't preprocess include files. Should we?
        if( $_ =~ $rx{include} )
        {
            $config{include} ||= [];
            push @{ $config{include} }, $1;
            #warn('Found include '.$1);
        }

        elsif( $_ =~ $rx{block_opening} )
        {
            # Start with a new block without options
            unshift @block_stack,     $block_name;
            unshift @block_opt_stack, $block_opts;

            $block_opts = {};
            $block_name = $1;
            
            for( $block_name )         # Trim block name
            {
                s/\s*$//;
                s/^\s*//;
            }

            #warn('Opening block `'.$block_name.'\'');
        }

        #
        # In this way (elsif), we're not going to parse correctly the
        # configuration directives that are written on a single line
        # ex.: `forwarders { 192.168.0.1; };'
        #
        elsif( $_ =~ $rx{block_closing} )
        {

            # Special handling for bind "zone" blocks
            if( $block_name =~ $rx{zone} )
            {
                my $zone_name = $1;
                $config{zones}->{$zone_name} = { %$block_opts };
            }
            else
            {

                my $parent_block_opts = $block_opt_stack[0];

                # If we have only list values (ex.: `forwarders { a; b; c; };')
                if( @values && !%$block_opts )
                {
                    $parent_block_opts->{$block_name} = [ @values ];
                }

                # Or if we have values *and* options (ex.: `masters { a; type slave; };')
                elsif( @values && %$block_opts )
                {
                    $parent_block_opts->{$block_name}->{values}  = [ @values ];
                    $parent_block_opts->{$block_name}->{options} = { %$block_opts };
                }

                # ... only options
                else
                {
                    $parent_block_opts->{$block_name} = { %$block_opts };
                }

                # If block name is not defined, we don't want to add a "root" block key
                if( $block_stack[0] )
                {
                    $config{$block_stack[0]} = $parent_block_opts;
                }
            }

            # Closing current block. Shift up the stack of 1 block and proceed
            @values     = ();
            $block_name = shift @block_stack;
            $block_opts = shift @block_opt_stack;

            #warn('Closing block `'.$block_name.'\'');
        }

        # Single value on a line
        elsif( $_ =~ $rx{value} )
        {
            push @values, $1;
        }

        # Option with keyword + value
        elsif( $_ =~ $rx{option} )
        {
            #warn('Found option `'.$1.'\' with value `'.$2.'\'');
            my($opt_name, $opt_value) = ($1, $2);
            for( $opt_name, $opt_value )
            {
                s/^\s*//;
                s/\s*$//;
            }

            # Check how we can insert this option in the existing hash
            # FIXME use check_hash_and_assign() ?
            if( exists $block_opts->{$opt_name} ) {
                if( ref $block_opts->{$opt_name} eq 'ARRAY' ) {
                    push @{ $block_opts->{$opt_name} }, $opt_value;
                } else {
                    $block_opts->{$opt_name} = [ $block_opts->{$opt_name}, $opt_value ];
                }
            } else {
                $block_opts->{$opt_name} = $opt_value;
            }
        }
    }

    return \%config;
}

sub irssi_style { croak "irssi-style config not supported in this release" }

# BUG: These functions are too similar. How can they be unified?

sub colon_sep {

    my $file = shift;
    open IN, $file or die $!;
    my %config;
    while (<IN>) {
        next if /^\s*#/;
        /^\s*(.*?)\s*:\s*(.*)/ or next;
        my ($k, $v) = ($1, $2);
        my @v;
        if ($v =~ /:/) {
            @v =  split /:/, $v;
        } elsif ($v =~ /, /) {
            @v = split /\s*,\s*/, $v;
        } elsif ($v =~ / /) {
            @v = split /\s+/, $v;
        } elsif ($v =~ /,/) { # Order is important
            @v = split /\s*,\s*/, $v;
        } else {
            @v = $v;
        }
        check_hash_and_assign(\%config, $k, @v);
    }
    return \%config;
}

sub check_hash_and_assign {
    my ($c, $k, @v) = @_;
    if (exists $c->{$k} and !ref $c->{$k}) {
        $c->{$k} = [$c->{$k}];
    }

    if (grep /=/, @v) { # Bugger, it's really a hash
        for (@v) {
            my ($subkey, $subvalue);
            if (/(.*)=(.*)/) { ($subkey, $subvalue) = ($1,$2); }
            else { $subkey = $1; $subvalue = 1; }

            if (exists $c->{$k} and ref $c->{$k} ne "HASH") {
                # Can we find a hash in here?
                my $h=undef;
                for (@{$c->{$k}}) {
                    last if ref ($h = $_) eq "hash";
                }
                if ($h) { $h->{$subkey} = $subvalue; }
                else { push @{$c->{$k}}, { $subkey => $subvalue } }
            } else {
                $c->{$k}{$subkey} = $subvalue;
            }
        }
    } elsif (@v == 1) {
        if (exists $c->{$k}) {
            if (ref $c->{$k} eq "HASH") { $c->{$k}{$v[0]} = 1; }
            else {push @{$c->{$k}}, @v}
        } else { $c->{$k} = $v[0]; }
    } else {
        if (exists $c->{$k}) {
            if (ref $c->{$k} eq "HASH") { $c->{$k}{$_} = 1 for @v }
            else {push @{$c->{$k}}, @v }
        }
        else { $c->{$k} = [@v]; }
    }
}


sub equal_sep {
    my $file = shift;
    open IN, $file or die $!;
    my %config;
    while (<IN>) {
        next if /^\s*#/;
        /^\s*(.*?)\s*=\s*(.*)\s*$/ or next;
        my ($k, $v) = ($1, $2);
        my @v;
        if ($v=~ /,/) {
            $config{$k} = [ split /\s*,\s*/, $v ];
        } elsif ($v =~ / /) { # XXX: Foo = "Bar baz"
            $config{$k} = [ split /\s+/, $v ];
        } else {
            $config{$k} = $v;
        }
    }

    return \%config;
}

sub space_sep {
    my $file = shift;
    open IN, $file or die $!;
    my %config;
    while (<IN>) {
        next if /^\s*#/;
        /\s*(\S+)\s+(.*)/ or next;
        my ($k, $v) = ($1, $2);
        my @v;
        if ($v=~ /,/) {
            @v = split /\s*,\s*/, $v;
        } elsif ($v =~ / /) { # XXX: Foo = "Bar baz"
            @v = split /\s+/, $v;
        } else {
            @v = $v;
        }
        check_hash_and_assign(\%config, $k, @v);
    }
    return \%config;

}

1;
__END__
# Below is stub documentation for your module. You better edit it!

=head1 NAME

Config::Auto - Magical config file parser

=head1 SYNOPSIS

  use Config::Auto;

  # Not very magical at all.
  my $config = Config::Auto::parse("myprogram.conf", format => "colon");

  # Considerably more magical.
  my $config = Config::Auto::parse("myprogram.conf");

  # Highly magical.
  my $config = Config::Auto::parse();

=head1 DESCRIPTION

This module was written after having to write Yet Another Config File Parser
for some variety of colon-separated config. I decided "never again".

When you call C<Config::Auto::parse> with no arguments, we first look at
C<$0> to determine the program's name. Let's assume that's C<snerk>. We
look for the following files:

    snerkconfig
    ~/snerkconfig
    /etc/snerkconfig
    /usr/local/etc/snerkconfig
    snerk.config
    ~/snerk.config
    /etc/snerk.config
    /usr/local/etc/snerk.config
    snerkrc
    ~/snerkrc
    /etc/snerkrc
    /usr/local/etc/snerkrc
    .snerkrc
    ~/.snerkrc
    /etc/.snerkrc
    /usr/local/etc/.snerkrc

Additional search paths can be specified with the C<paths> option.

We take the first one we find, and examine it to determine what format
it's in. The algorithm used is a heuristic "which is a fancy way of
saying that it doesn't work." (Mark Dominus.) We know about colon
separated, space separated, equals separated, XML, Perl code, Windows
INI, BIND9 and irssi style config files. If it chooses the wrong one,
you can force it with the C<format> option.

If you don't want it ever to detect and execute config files which are made
up of Perl code, set C<$Config::Auto::DisablePerl = 1>.

When using the perl format, your configuration file will be eval'd. This will
cause taint errors. To avoid these warnings, set C<$Config::Auto::Untaint = 1>.

When using the perl format, your configuration file will be eval'd using
do(file). This will cause taint errors if the filename is not untainted. To
avoid these warnings, set C<$Config::Auto::Untaint = 1>. This setting will not
untaint the data in your configuration file and should only be used if you
trust the source of the filename.

Then the file is parsed and a data structure is returned. Since we're
working magic, we have to do the best we can under the circumstances -
"You rush a miracle man, you get rotten miracles." (Miracle Max) So
there are no guarantees about the structure that's returned. If you have
a fairly regular config file format, you'll get a regular data
structure back. If your config file is confusing, so will the return
structure be. Isn't life tragic?

Here's what we make of some common Unix config files:

F</etc/resolv.conf>:

    $VAR1 = {
          'nameserver' => [ '163.1.2.1', '129.67.1.1', '129.67.1.180' ],
          'search' => [ 'oucs.ox.ac.uk', 'ox.ac.uk' ]
        };

F</etc/passwd>:

    $VAR1 = {
          'root' => [ 'x', '0', '0', 'root', '/root', '/bin/bash' ],
          ...
        };

F</etc/gpm.conf>:

    $VAR1 = {
          'append' => '""',
          'responsiveness' => '',
          'device' => '/dev/psaux',
          'type' => 'ps2',
          'repeat_type' => 'ms3'
        };

F</etc/nsswitch.conf>:

    $VAR1 = {
          'netgroup' => 'nis',
          'passwd' => 'compat',
          'hosts' => [ 'files', 'dns' ],
          ...
    };

=head1 PARAMETERS

Although C<Config::Auto> is at its most magical when called with no parameters,
its behavior can be reined in by use of one or two arguments.

If a filename is passed as the first argument to C<parse>, the same paths are
checked, but C<Config::Auto> will look for a file with the passed name instead
of the C<$0>-based names.

 use Config::Auto;

 my $config = Config::Auto::parse("obscure.conf");

The above call will cause C<Config::Auto> to look for:

 obscure.conf
 ~/obscure.conf
 /etc/obscure.conf

Parameters after the first are named.

=head2 C<format> 

forces C<Config::Auto> to interpret the contents of the
configuration file in the given format without trying to guess.
 
=head2 C<path>

add additional directories to the search paths. The current directory
is searched first, then the paths specified with the path parameter.
C<path> can either be a scalar or a reference to an array of paths to check.

=head2 Formats

C<Config::Auto> recognizes the following formats:

=over 4

=item * perl    => perl code

=item * colon   => colon separated (e.g., key:value)

=item * space   => space separated (e.g., key value)

=item * equal   => equal separated (e.g., key=value)

=item * bind    => bind style (not available)

=item * irssi   => irssi style (not available)

=item * xml     => xml (via XML::Simple)

=item * ini     => .ini format (via Config::IniFiles)

=item * list    => list (e.g., ??)

=item * yaml    => yaml (via YAML.pm)

=back

=head1 TROUBLESHOOTING

=over 4

=item When using a Perl config file, the configuration is borked

Give C<Config::Auto> more hints (e.g., add #!/usr/bin/perl to beginning of
file) or indicate the format in the parse() command.

=back

=head1 TODO

Better BIND9 file format parser.
irssi file format parser currently doesn't exist. It would be
good to add support for "mutt" and "vim" style "set"-based RCs.

=head1 AUTHOR

This module by Jos Boumans, C<kane@cpan.org>.

=head1 LICENSE

This module is
copyright (c) 2003-2006 Jos Boumans E<lt>kane@cpan.orgE<gt>.
All rights reserved.

This library is free software;
you may redistribute and/or modify it under the same
terms as Perl itself.

=cut
